src/Shortcodes/ImageShortcodeProvider.php

```diff
@@ -112,11 +112,14 @@ public static function handle_shortcode($args, $content, $parser, $shortcode, $e
 
         // Clean out any empty attributes (aside from alt) and anything not whitelisted
         $whitelist = static::config()->get('attribute_whitelist');
-        $attrs = array_filter($attrs ?? [], function ($v, $k) use ($whitelist) {
-            return in_array($k, $whitelist) && (strlen(trim($v ?? '')) || $k === 'alt');
-        }, ARRAY_FILTER_USE_BOTH);
+        foreach ($attrs as $key => $value) {
+            if (in_array($key, $whitelist) && (strlen(trim($value ?? '')) || $key === 'alt')) {
+                $manipulatedRecord = $manipulatedRecord->setAttribute($key, $value);
+            }
+        }
 
-        $markup = self::createImageTag($attrs);
+        // We're calling renderWith() with an explicit template in case someone wants to use a custom template
+        $markup = $manipulatedRecord->renderWith(self::class . '_Image');
```

Comment:
We still need to call `createImageTag()` because people may have subclassed the shortcode provider, replaced the original one with theirs, and updated the `createImageTag()` method. If we stop calling it, we throw away their customisation which is a breaking change.

Instead, update the implementation of `createImageTag()` to return markup which is generated via the template. Alternatively update it to return an empty string. Then we can call it and if it returns a truthy value we know it's a custom implementation and we should use that instead of the template.

Reply:
On the broader point, I don't think that think semantic versionning means we will always call the same methods in the same order.

On a more narrow point, because the `createImageTag` call uses `self` rather than `static`, if you subclass `ImageShortcodeProvider` and override `createImageTag`, calling `handle_shortcode` on your subclass will still end up calling `ImageShortcodeProvider::createImageTag()`

Reply:
Ahh, I didn't catch that it uses self. I fully disagree with your reasoning for the first point but it's made moot by the use of self.